solar_crypto/transactions/builder/vote.py

```diff
@@ -1,38 +1,96 @@
+import re
 import typing
+from functools import cmp_to_key
+from math import trunc
 
-from solar_crypto.constants import TRANSACTION_VOTE
-from solar_crypto.identity.address import address_from_passphrase
+from solar_crypto.constants import SOLAR_TRANSACTION_VOTE, TRANSACTION_TYPE_GROUP
 from solar_crypto.transactions.builder.base import BaseTransactionBuilder
 
 
 class Vote(BaseTransactionBuilder):
 
-    transaction_type = TRANSACTION_VOTE
+    transaction_type = SOLAR_TRANSACTION_VOTE
+    typeGroup = TRANSACTION_TYPE_GROUP.SOLAR.value
 
-    def __init__(self, vote=None, fee=None):
-        """Create a second signature registration transaction
-
-        Args:
-            vote (str): address of a delegate you want to vote
-            fee (int, optional): fee used for the transaction (default is already set)
-        """
+    def __init__(self):
         super().__init__()
 
         self.transaction.asset["votes"] = []
-        if vote:
-            self.transaction.asset["votes"].append(vote)
 
-        if fee:
-            self.transaction.fee = fee
-
-    def set_votes(self, votes: typing.List[str]):
-        """Set votes/unvotes
+    def set_votes(
+        self,
+        votes: typing.Union[typing.List[str], typing.Dict[str, typing.Union[int, float]]] = dict,
+    ):
+        """Set votes
 
         Args:
-            votes (List[str]): list of votes
+            votes
         """
+        vote_object: typing.Dict[str, typing.Union[float, int]] = {}
+
+        if isinstance(votes, list):
+            vote_list = filter(lambda vote: not vote.startswith("-"), votes)
+            vote_list = list(map(lambda vote: vote[1:], vote_list))
+
+            if len(vote_list) == 0:
+                self.transaction.asset["votes"] = {}
+                return
+
```

Comment:
Where does core do this validation? I'm not seeing it in the builder or serializer.

Reply:
https://github.com/Solar-network/core/blob/7c4c74f91362cd33f083131d0bf97f0627e4f729/packages/crypto/src/transactions/builders/transactions/solar/vote.ts#L61

https://github.com/Solar-network/core/blob/7c4c74f91362cd33f083131d0bf97f0627e4f729/packages/crypto/src/transactions/builders/transactions/transaction.ts#L160-L166

https://github.com/Solar-network/core/blob/7c4c74f91362cd33f083131d0bf97f0627e4f729/packages/crypto/src/transactions/types/schemas.ts#L157-L183

https://github.com/Solar-network/core/blob/7c4c74f91362cd33f083131d0bf97f0627e4f729/packages/crypto/src/validation/index.ts#L17-L55

Reply:
Ah, I misread (too early 😅). You specifically ask about where Core does validation for > 53 rather than generally where the validation happens.

It used to do it in crypto, now it does it here: https://github.com/Solar-network/core/blob/7c4c74f91362cd33f083131d0bf97f0627e4f729/packages/core-transactions/src/handlers/solar/vote.ts#L54-L59

Since it's possible the number of delegates could theoretically change in future, so it's not something for crypto to handle anymore.

Reply:
It used to do it in crypto, now it does it here: https://github.com/Solar-network/core/blob/7c4c74f91362cd33f083131d0bf97f0627e4f729/packages/core-transactions/src/handlers/solar/vote.ts#L54-L59

It was moved there since it's possible the number of delegates could theoretically change in future, so it's not something for crypto to handle anymore.